### DISCUSSION TITLE
Extend with JavaScript: how to save a photo encoded in base64 string

### DISCUSSION THREAD

**Discussion by benny856694** [pocketbase/pocketbase#4007] [Q&A] [answered]
I want to save the photo to a collection, and the photo is defined as file type, {"name": "photo", "type": "file"}. 
Thanks for the great project.

**Comment by ganigeorgiev** [comment]
How the base64 string is submitted?

Why does in need to be a base64 string in the first place?

**Reply by benny856694** [reply]
Posted in json

**Reply by jonshipman** [reply]
I mean, then you save it as a string in the database. If you want to upload a file, I believe that's a form-data request.
See https://pocketbase.io/docs/files-handling/

**Reply by benny856694** [reply]
I can do that, just concerned about the performance if too many.

**Reply by pondrax** [reply]
I'm interested in understanding this as well. For instance, I'm currently taking screenshots of a website using the PageSpeed API with a cron command. The output is in the form of a base64 image string. If I save it as a string, the network request will be blocked until all the string records are loaded. However, if I save it as a file, each network request will be split into a record collection and a file request.

The documentation for uploading only shows $filesystem.fileFromPath, but in JSVM, there is also filesystem.newFileFromUrl. See  https://pocketbase.io/docs/js-records/#create-new-record-with-data-validations and https://pocketbase.io/jsvm/interfaces/filesystem.newFileFromUrl.html

**Reply by jonshipman** [reply]
You can always store base64 data in a relation table

**Reply by pondrax** [reply, marked as answer]
finally got it, using $filesystem.fileFromBytes(). i have to convert base64 to uint8 first tho

**Reply by benny856694** [reply]
wow, thanks a lot, you saved my day.

**Reply by benny856694** [reply]
Is this function newly added?

**Reply by marquaye** [reply]
@pondrax How did you manage to convert the base64 string to uint8? It seems goja doesn't support textDecoder nor atop to do the conversion.

**Reply by marquaye** [reply]
For anyone having the same Issue. 
While I agree with @ganigeorgiev that using a different format than base64 is the best way, in my case I had to deal with base64 encoded pictures.

I found a pure js implementation for atob and used the `$filesystem.fileFromBytes` to create files.
Here is a quick implementation:

                function base64ToByteArray(base64String) { 
                const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';
                const byteArray = new Uint8Array(Math.ceil(base64String.length * 0.75));
            
                let byteIndex = 0;
                let charIndex = 0;
            
                while (charIndex < base64String.length) {
                    let enc1 = chars.indexOf(base64String.charAt(charIndex++));
                    let enc2 = chars.indexOf(base64String.charAt(charIndex++));
                    let enc3 = chars.indexOf(base64String.charAt(charIndex++));
                    let enc4 = chars.indexOf(base64String.charAt(charIndex++));
            
                    let bits24 = (enc1 << 18) | (enc2 << 12) | (enc3 << 6) | enc4;
            
                    let byte1 = (bits24 >> 16) & 0xff;
                    let byte2 = (bits24 >> 8) & 0xff;
                    let byte3 = bits24 & 0xff;
            
                    byteArray[byteIndex++] = byte1;
            
                    if (enc3 !== 64) {
                        byteArray[byteIndex++] = byte2;
                        if (enc4 !== 64) {
                            byteArray[byteIndex++] = byte3;
                        }
                    }
                }
            
                return byteArray.subarray(0, byteIndex);
            }

After the conversion I used the `RecordUpsertForm` to update the coresponding record in my collection:
            
            const byteArray = base64ToByteArray(Image);
            const file = $filesystem.fileFromBytes(byteArray, "Picture_2341.jpg");
                    
            const record = new Record(collection)
            const form = new RecordUpsertForm($app, record)
            form.addFiles("picture", file)
            form.submit()

There is a discussion in [goja](https://github.com/dop251/goja/discussions/541) about adding the atob function, but it seems that it has been stalled for a year. I think it would be nice to have an atob and btop function to avoid the hassle when converting base64 strings and the proposed solution mentioned in the discussion seems very easy to implement, but as I am in my early days with go I can not help with implementation.

**Reply by jonshipman** [reply]
Beats my solution of making a REST request to the frontend 🤣

**Comment by Pwea** [comment]
For anyone having trouble with @marquaye's solution.
I made one that should work for the newer versions of Pocketbase.
My Pocketbase version: 0.24.3

Here's a demo implementation:
```javascript
const base64ToByteArray = (base64String) => {
    const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
  
    // Remove padding if present
    base64String = base64String.replace(/=/g, '');
  
    // Adjust length to be divisible by 4 by adding "A" (neutral character in Base64)
    while (base64String.length % 4 !== 0) {
      base64String += 'A';
    }
  
    const byteArray = new Uint8Array((base64String.length * 3) / 4);
  
    let byteIndex = 0;
    let charIndex = 0;
  
    while (charIndex < base64String.length) {
      const enc1 = chars.indexOf(base64String.charAt(charIndex++));
      const enc2 = chars.indexOf(base64String.charAt(charIndex++));
      const enc3 = chars.indexOf(base64String.charAt(charIndex++));
      const enc4 = chars.indexOf(base64String.charAt(charIndex++));
  
      if (enc1 === -1 || enc2 === -1 || enc3 === -1 || enc4 === -1) {
        throw new Error('Invalid character in Base64 string');
      }
  
      const bits24 = (enc1 << 18) | (enc2 << 12) | (enc3 << 6) | enc4;
  
      const byte1 = (bits24 >> 16) & 0xff;
      const byte2 = (bits24 >> 8) & 0xff;
      const byte3 = bits24 & 0xff;
  
      byteArray[byteIndex++] = byte1;
      if (enc3 !== 64) {
        byteArray[byteIndex++] = byte2;
        if (enc4 !== 64) {
          byteArray[byteIndex++] = byte3;
        }
      }
    }
  
    return byteArray.subarray(0, byteIndex);
};
```

And then how to use it:
Using this as reference: https://pocketbase.io/docs/js-records/#create-new-record-programmatically
When saving the images, you can either give one or a list of multiple.
```javascript
const collection = $app.findCollectionByNameOrId("objects")

const record = new Record(collection, {
    "title": "test",
    "description": "testing",
    "user": "c96r6g7015adxsv",
    "rarity": 4
})

let imagesAsBytes = [];

for (let i = 0; i < images.length; i++) {
    const byteArray = base64ToByteArray(data.images[i].data);
    const file = $filesystem.fileFromBytes(byteArray, data.images[i].filename);

    imagesAsBytes.push(file)
}

record.set("pictures", imagesAsBytes)
$app.save(record);
```

This implementation worked for me on my Pocketbase version.

I hope this helps!

**Reply by ganigeorgiev** [reply]
Many of the above replies contain only _"I want/I need"_ and no clear explanation and reason WHY this is even needed in the first place.

I'll strongly discourage accepting files as base64 string or storing them into the database as such.

It not only will affect the query performance and memory usage but you also risking various mimetype/size related security vulnerabilities if not used with proper care.
